go/tools/gazelle/packages/package.go

```diff
@@ -136,3 +137,171 @@ func (ts *PlatformStrings) firstGoFile() string {
+// be added to any target (for example, .txt files).
+func (p *Package) addFile(info fileInfo, cgo bool, buildTags map[string]bool, platforms PlatformConstraints) error {
+	var t *Target
+	if info.isXTest {
+		if info.isCgo {
+			return fmt.Errorf("%s: use of cgo in test not supported", info.path)
+		}
+		t = &p.XTest
+	} else if info.isTest {
+		if info.isCgo {
+			return fmt.Errorf("%s: use of cgo in test not supported", info.path)
+		}
+		t = &p.Test
+	} else if info.isCgo || cgo && (info.category == cExt || info.category == hExt || info.category == csExt) {
+		t = &p.CgoLibrary
+	} else if info.category == goExt || info.category == sExt || !cgo && info.category == hExt {
+		t = &p.Library
+	}
+	if t != nil {
+		t.addFile(info, buildTags, platforms)
+	}
+	return nil
+}
+
+func (t *Target) addFile(info fileInfo, buildTags map[string]bool, platforms PlatformConstraints) {
+	if !info.hasConstraints() || info.checkConstraints(buildTags) {
+		t.Sources.addGenericStrings(info.name)
+		t.Imports.addGenericStrings(info.imports...)
+		t.COpts.addGenericOpts(platforms, info.copts)
+		t.CLinkOpts.addGenericOpts(platforms, info.clinkopts)
+		return
+	}
+
+	for name, tags := range platforms {
+		if info.checkConstraints(tags) {
+			t.Sources.addPlatformStrings(name, info.name)
+			t.Imports.addPlatformStrings(name, info.imports...)
+			t.COpts.addTaggedOpts(name, info.copts, tags)
+			t.CLinkOpts.addTaggedOpts(name, info.clinkopts, tags)
+		}
+	}
+}
+
+func (ps *PlatformStrings) addGenericStrings(ss ...string) {
+	ps.Generic = append(ps.Generic, ss...)
+}
+
+func (ps *PlatformStrings) addGenericOpts(platforms PlatformConstraints, opts []taggedOpts) {
+	for _, t := range opts {
+		if t.tags == "" {
+			ps.Generic = append(ps.Generic, t.opts...)
+			continue
+		}
+
+		for name, tags := range platforms {
+			if checkTags(t.tags, tags) {
+				ps.Platform[name] = append(ps.Platform[name], t.opts...)
+			}
+		}
+	}
+}
+
+func (ps *PlatformStrings) addPlatformStrings(name string, ss ...string) {
+	if ps.Platform == nil {
+		ps.Platform = make(map[string][]string)
+	}
+	ps.Platform[name] = append(ps.Platform[name], ss...)
+}
+
+func (ps *PlatformStrings) addTaggedOpts(name string, opts []taggedOpts, tags map[string]bool) {
+	if ps.Platform == nil {
+		ps.Platform = make(map[string][]string)
+	}
+	for _, t := range opts {
+		if t.tags == "" || checkTags(t.tags, tags) {
+			ps.Platform[name] = append(ps.Platform[name], t.opts...)
+		}
+	}
+}
+
+// Clean sorts and de-duplicates PlatformStrings. It also removes any
+// strings from platform-specific lists that also appear in the generic list.
+// This is useful for imports.
+func (ps *PlatformStrings) Clean() {
+	sort.Strings(ps.Generic)
+	ps.Generic = uniq(ps.Generic)
+
+	genSet := make(map[string]bool)
+	for _, s := range ps.Generic {
+		genSet[s] = true
+	}
+
+	if ps.Platform != nil {
+		for n, ss := range ps.Platform {
+			ss = remove(ss, genSet)
```

Comment:
I feel like this whole bit would be much easier to understand (and probably faster) if you just converted it to a map, then twiddled the bools (possibly converting back to a list at the end if you really really need a list rather than a map), rather than sorting and uniquing